### DISCUSSION TITLE
regression in consistent hashing plugin after #5077

### DISCUSSION THREAD

**Discussion by lazedo** [rabbitmq/rabbitmq-server#11671] [Other] [answered]
we work a lot with consistent hashing and noticed a regression in behaviour after #5077 

when consistency is configured with a `header` and that `header` is not set on the published msg, a crash in `rabbit_exchange_type_consistent_hash.erl` , this is different from previous versions. if this is the new desired behaviour, how can we prevent clients from publishing without the `header`? can the msg be rejected ?

```
value_to_hash({header, Header}, Msg0) ->
    maps:get(Header, mc:routing_headers(Msg0, [x_headers]));
```

should be changed to below to restore behaviour 

```
value_to_hash({header, Header}, Msg0) ->
    maps:get(Header, mc:routing_headers(Msg0, [x_headers]), undefined);
```

**Comment by lukebakken** [comment]
Hello, would you please provide the full stack trace you see with this crash?

**Comment by hugh-zen** [comment]
The same issue is happening to me and i've reproduced it easily by creating a consistent hashing exchange with: hash-header = hash_id

Then publish a message without the header set. 

> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0> Connection (<0.27400.0>) closing: internal error in channel (<0.27408.0>): {{badkey,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                              <<"hash_id">>},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                             [{erlang,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               map_get,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               [<<"hash_id">>,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                #{}],
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               [{error_info,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 #{module =>
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                    erl_erts_errors}}]},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                              {rabbit_exchange_type_consistent_hash,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               value_to_hash,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               2,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               [{file,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 "rabbit_exchange_type_consistent_hash.erl"},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                {line,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 265}]},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                              {rabbit_exchange_type_consistent_hash,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               route,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               3,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               [{file,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 "rabbit_exchange_type_consistent_hash.erl"},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                {line,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 71}]},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                              {rabbit_exchange,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               route1,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               4,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               [{file,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 "rabbit_exchange.erl"},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                {line,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 389}]},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                              {rabbit_exchange,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               route,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               3,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               [{file,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 "rabbit_exchange.erl"},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                {line,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 369}]},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                              {rabbit_channel,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               handle_method,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               3,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               [{file,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 "rabbit_channel.erl"},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                {line,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 1281}]},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                              {rabbit_channel,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               handle_cast,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               2,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               [{file,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 "rabbit_channel.erl"},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                {line,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 659}]},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                              {gen_server2,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               handle_msg,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               2,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                               [{file,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 "gen_server2.erl"},
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                {line,
> 2024-07-23 13:42:13.476706+00:00 [warning] <0.27400.0>                                                                                 1056}]}]}
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>     supervisor: {<0.27406.0>,amqp_channel_sup}
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>     errorContext: shutdown
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>     reason: reached_max_restart_intensity
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>     offender: [{pid,<0.27408.0>},
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>                {id,channel},
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>                {mfargs,{amqp_channel,start_link,
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>                                      [direct,<0.27400.0>,1,<0.27407.0>,
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>                                       {<<"<rabbit@003f76197cf0.1721731945.27400.0>">>,
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>                                        1}]}},
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>                {restart_type,transient},
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>                {significant,true},
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>                {shutdown,300000},
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>                {child_type,worker}]
> 2024-07-23 13:42:13.477116+00:00 [error] <0.27406.0>
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>     supervisor: {<0.27410.0>,rabbit_channel_sup}
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>     errorContext: child_terminated
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>     reason: {{badkey,<<"hash_id">>},
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>              [{erlang,map_get,
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>                   [<<"hash_id">>,#{}],
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>                   [{error_info,#{module => erl_erts_errors}}]},
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>               {rabbit_exchange_type_consistent_hash,value_to_hash,2,
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>                   [{file,"rabbit_exchange_type_consistent_hash.erl"},
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>                    {line,265}]},
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>               {rabbit_exchange_type_consistent_hash,route,3,
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>                   [{file,"rabbit_exchange_type_consistent_hash.erl"},
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>                    {line,71}]},
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>               {rabbit_exchange,route1,4,
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>                   [{file,"rabbit_exchange.erl"},{line,389}]},
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>               {rabbit_exchange,route,3,
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>                   [{file,"rabbit_exchange.erl"},{line,369}]},
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>               {rabbit_channel,handle_method,3,
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>                   [{file,"rabbit_channel.erl"},{line,1281}]},
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>               {rabbit_channel,handle_cast,2,
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>                   [{file,"rabbit_channel.erl"},{line,659}]},
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>               {gen_server2,handle_msg,2,
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>                   [{file,"gen_server2.erl"},{line,1056}]}]}
> 2024-07-23 13:42:13.476551+00:00 [error] <0.27410.0>     offender: [{pid,<0.27412.0>},

**Comment by michaelklishin** [comment]
This is just a rarely used way of using the plugin. Please put together an executable way to reproduce (as a public repo on GitHub).

**Comment by jeremy-smith-maco** [comment]
I have also encountered this issue. I have created a reproduction case using a simple python script taken from the consistent hashing example. You can see the original line on 28 I replaced with an empty headers object on line 30. https://gist.github.com/jeremy-smith-maco/1dc7eeee7ae189678e98de93fdbebf8f

**Comment by ansd** [comment, marked as answer]
Fixed in https://github.com/rabbitmq/rabbitmq-server/pull/11808
